Lookup cache

sebdalgarno opened the pull request:
generate and cache lookup tables for ShinyRems app
fixes #41

sebdalgarno committed:
Merge pull request #2 from bcgov/master

sync forks

sebdalgarno committed:
generate and cache lookup for shinyrems app

sebdalgarno committed:
Merge branch 'master' into lookup_cache

ateucher commented:
@sebdalgarno is this ready for review?

sebdalgarno commented:
yes please (I don't have privilege to assign reviewers on this repo)

ateucher commented on R/utils-shinyrems.R:
```diff
@@ -0,0 +1,80 @@
+#'
+#' \code{"EMS_ID", "MONITORING_LOCATION", "PERMIT", "PARAMETER_CODE", "PARAMETER",
+#' "LONGITUDE" "LATITUDE", "FROM_DATE", "TO_DATE"}
+#' @export
```
Suggest adding `internal` keyword to documentation. This means that while you can still get the help for the function using `?get_ems_lookup` it won't be added to the package documentation index, which I would prefer since it's not meant for general use.

```suggestion
#' @export
#' @keywords internal
```

ateucher commented on R/utils-shinyrems.R:
```diff
@@ -0,0 +1,80 @@
+    message("Creating and caching lookup table ...")
+    data <- try(._remsCache_$get(which), silent = TRUE)
+    if(inherits(data, "try-error"))
+      stop(which, " dataset must be cached before lookup table can be created. Run get_ems_data().")
```
Rather than a `try` here, can you check if `which` exists with `._remsCache_$exists(which)`, it would be a bit cleaner I think

ateucher commented on R/utils-shinyrems.R:
```diff
@@ -0,0 +1,80 @@
+    if (lup_cache_date < data_cache_date){
+      update <- TRUE
+    }
+  }
```
Do you need a final `else` here to catch the case where `which_exists && !._remsCache_$exists("cache_dates")`? If not, is it safe to just use an `else` on line 27, assuming that if `which_exists` there will also be a `cache_dates`?

ateucher replied:
In fact, `get_cache_date()` checks if it exists already...

sebdalgarno replied:
yes good point. I have removed the else statement entirely. if cache_dates don't exist, both will return -Inf and -Inf < -Inf will evaluate to FALSE. Note, this code was basically copied from get_ems_data - might be worth reviewing that as well.

ateucher replied:
Ok, thanks, I will try to have a look. In that case I think all of line 25-35 can be replaced with:

```r
lup_cache_date <- get_cache_date(which_lup)
data_cache_date <- get_cache_date(which)

update <- !which_exists || lup_cache_date < data_cache_date
```

sebdalgarno replied:
very nice! will make that change

ateucher commented on R/utils-shinyrems.R:
```diff
@@ -0,0 +1,80 @@
+lookup_from_cache <- function(which){
+  ._remsCache_$get(which)
+}
+
```
Is this necessary? It's only used once as far as I can see, and using ` ._remsCache_$get(which)` directly has a clear enough intent I think

sebdalgarno replied:
not necessary - will remove

ateucher reviewed (changes requested):
Thanks @sebdalgarno looks good. A few comments, and I think a run of `styler` would be helpful re spacing etc.

sebdalgarno committed:
Update R/utils-shinyrems.R

Co-Authored-By: Andy Teucher <andy.teucher@gmail.com>

sebdalgarno committed:
replace try with ._remsCache_$exists

sebdalgarno committed:
remove else if statement and check for cache_dates

sebdalgarno committed:
remove lookup_from_cache fun

sebdalgarno committed:
run styler

sebdalgarno committed:
resolve merge conflict

Merge branch 'lookup_cache' of github.com:poissonconsulting/rems into lookup_cache

# Conflicts:
#	R/utils-shinyrems.R

sebdalgarno committed:
more elegant update

ateucher commented:
Thanks for the quick fixes! I think if you run `devtools::document()` we're good to go

sebdalgarno committed:
document changes
